### ISSUE TITLE
Validate unique rate limit configuration

### ISSUE THREAD

**Issue by ryanslade** [sourcegraph/sourcegraph#10026]
Currently, if internal rate limiters are configured for more than one external service that belong to the same code host, we pick the lowest limit. That is, the limiter that is most restrictive.

This will be documented, but we should also display a warning to the user if they configure more than one rate limiter for the same code host.

**Comment by ryanslade** [comment]
The simplest way to implement this is to populate the warning here and display it on save:
https://github.com/sourcegraph/sourcegraph/blob/1e40756fceda328c4b3b66f8d386326dc21ddbf4/cmd/frontend/graphqlbackend/external_service.go#L21

However, we already do this in other places, for example:
https://github.com/sourcegraph/sourcegraph/blob/1e40756fceda328c4b3b66f8d386326dc21ddbf4/cmd/frontend/graphqlbackend/external_services.go#L53

There are a couple of problems with this:

1. How should we format when more than one warning appears in the `warning` string?
2. When config is updated from the frontend we actually make an http requests to repo-updater to validate the config so the warning would need to be passed back from there.

Considering the above, do we think this change is worth making or is documenting the behaviour enough?

**Comment by slimsag** [comment]
Most of your pain/challenges here come from a bad choice we made a while ago to make external services use configuration separate from the rest of the entire app's architecture (site config).

Luckily, someone has done the work for us already to sideload/inject external services problems into the same reporting mechanisms we use for site configuration so it's mostly the same for how you should implement this.

In short, you just need to add an external service validator here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@8dccda8f08f7fbad697bda0b67c54fe8ac38a60b/-/blob/cmd/frontend/graphqlbackend/site_alerts.go#L127-134 and it will then show up as a banner instead of "just on save"

**Comment by ryanslade** [comment]
Great, thanks for the pointer, that should make things a lot easier

**Comment by uwedeportivo** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.16 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by mrnugget** [comment]
> Most of your pain/challenges here come from a bad choice we made a while ago to make external services use configuration separate from the rest of the entire app's architecture (site config).

Can you expand on this? I'm curious as to what the road-to-taken would look like.

---

As I wrote here, https://github.com/sourcegraph/sourcegraph/pull/10661#issuecomment-628620955, I think that this is another validation we can do at _write_ time and immediately let the user know about the problem. We already have validations in place and could just extend those to produce another error.

That would be a relatively straight forward change and it would keep all the validations in a single place, instead of one producing an asynchronous site-alert and the others producing errors when creating/updating an external service.

Afterwards we could think about whether to change the way we do validations. But for now I wouldn't introduce another place where logic about external service validation is kept.
